spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+
+<div class="example">
+  `a.example` randomly assigns users to groups in a way that is consistent cross-site.
 
-<span class=XXX>TODO: Finish adding interfaces.</span>
-<span class=XXX>TODO: Add algorithms for APIs.</span>
-<span class=XXX>TODO: Add text, including fleshing out intro.</span>
+  Inside an `a.example` iframe:
+  <pre class="lang-js">
+    function generateSeed() { … }
+    await window.sharedStorage.worklet.addModule('experiment.js');
+
+    // Only write a cross-site seed to a.example's storage if there isn't one yet.
+    window.sharedStorage.set('seed', generateSeed(), { ignoreIfPresent: true });
+
+    let fencedFrameConfig = await window.sharedStorage.selectURL(
+      'select-url-for-experiment',
+      [
+        {url: "blob:https://a.example/123…", reportingMetadata: {"click": "https://report.example/1..."}},
+        {url: "blob:https://b.example/abc…", reportingMetadata: {"click": "https://report.example/a..."}},
+        {url: "blob:https://c.example/789…"}
+      ],
+      { data: { name: 'experimentA' } }
+    );
+
+    // Assumes that the fenced frame 'my-fenced-frame' has already been attached.
+    document.getElementById('my-fenced-frame').config = fencedFrameConfig;
+  </pre>
+
+  inside the `experiment.js` worklet script:
+  <pre class="lang-js">
+    class SelectURLOperation {
+      hash(experimentName, seed) { … }
+
+      async run(data, urls) {
+        const seed = await this.sharedStorage.get('seed');
+        return hash(data.name, seed) % urls.length;
+      }
+    }
+    register('select-url-for-experiment', SelectURLOperation);
+  </pre>
+</div>
+
+The {{SharedStorageWorklet}} Interface {#worklet}
+=================================================
+The {{SharedStorageWorklet}} object allows developers to supply [=module scripts=] to process [=Shared Storage=] data and then output the result through one or more of the output gates. Currently there are two output gates, the [=private aggregation=] output gate and the {{WindowSharedStorage/selectURL()|URL-selection}} output gate.
 
-Shared Storage DOM Interface {#dom-interface}
-=============================================
 <xmp class='idl'>
+  [Exposed=(Window)]
+  interface SharedStorageWorklet : Worklet {
+  };
+</xmp>
+
+  ## The {{SharedStorageWorkletGlobalScope}} ## {#global-scope}
+
+  The {{SharedStorageWorklet}}'s [=worklet global scope type=] is {{SharedStorageWorkletGlobalScope}}. 
+
+  <xmp class='idl'>
+    [Exposed=SharedStorageWorklet, Global=SharedStorageWorklet]
+    interface SharedStorageWorkletGlobalScope : WorkletGlobalScope {
+      undefined register(DOMString name,
+                         SharedStorageOperationConstructor operationCtor);
+    };
+
+    callback SharedStorageOperationConstructor = SharedStorageOperation(object options);
```

Comment:
How is the `object options` argument to the SharedStorageOperation used?  Can that be a more specific type than "object"?